librz/il/vm_layer.c

```diff
@@ -128,16 +128,21 @@ static void free_bind_var_val(HtPPKv *kv) {
  * initiate an empty VM
  * \param vm RzILVM, pointer to an empty VM
  * \param start_addr ut64, initiation pc address
- * \param addr_size ut32, size of the address in VM
- * \param data_size ut32, size of the minimal data unit in VM
+ * \param addr_size  ut32, size of the address in VM
+ * \param ro_memory  RzBuffer, read only memory to use for reads in the VM
  */
-RZ_API bool rz_il_vm_init(RzILVM *vm, ut64 start_addr, ut32 addr_size, ut32 data_size) {
-	vm->addr_size = addr_size;
-	vm->data_size = data_size;
+RZ_API bool rz_il_vm_init(RzILVM *vm, ut64 start_addr, ut32 addr_size, RzBuffer *ro_memory, bool big_endian) {
+	rz_return_val_if_fail(vm && ro_memory, false);
 
 	rz_pvector_init(&vm->vm_global_variable_list, (RzPVectorFree)rz_il_variable_free);
 	rz_pvector_init(&vm->vm_local_variable_list, (RzPVectorFree)rz_il_variable_free);
-	rz_pvector_init(&vm->vm_memory, (RzPVectorFree)rz_il_mem_free);
+
+	vm->vm_memory = rz_buf_new_sparse_overlay(ro_memory, RZ_BUF_SPARSE_WRITE_MODE_SPARSE);
```

Comment:
In the long run, using a sparse overlay buffer should not be hardcoded here since there are multiple use-cases of for example also just letting the vm persist everything in io directly. What is desired can be controlled by just supplying a buffer in the right form into this function.
But this is easy to refactor, so no blocker imo.